lucetc/src/function.rs

```diff
@@ -119,98 +227,147 @@ impl<'a> FuncInfo<'a> {
+                    //   i32.const 4   ; counter += 1
+                    //   i32.const -5  ; counter += 1
+                    //   i32.add       ; counter += 1
+                    //   block         ; flush counter (counter = 3 -> 0), flush here to avoid having
+                    //                 ;                                   accumulated count at the
+                    //                 ;                                   final `end`
+                    //     i32.const 4 ; counter += 1
+                    //     i32.add     ; counter += 1
+                    //   end           ; flush counter (counter = 2 -> 0)
+                    // end             ; flush counter (counter = 0 -> 0) and is a no-op
+                    // ```
+                    flush_counter(self, builder);
+                }
+                _ => { /* regular operation, do nothing */ }
             }
-            Operator::End => {
-                // We have to be really careful here to avoid violating a cranelift invariant:
-                // if the next operation is `End` as well, this end will have marked the block
-                // finished, and attempting to add instruction to update the instruction counter
-                // will cause a panic.
-                //
-                // We can avoid that case by ensuring instruction counts are flushed at the *entry*
-                // of any block-opening operation, so that at the exit the `End` will update the
-                // count by 0, the update is discarded, and we don't cause a panic.
-                flush_counter(self, builder);
-            }
-            _ => { /* regular operation, do nothing */ }
         }
 
         // finally, we might have to set up a new counter for a new scope, or fix up counts a bit
         match op {
+            Operator::Br { .. }
+            | Operator::BrTable { .. }
+            | Operator::Unreachable
+            | Operator::Return => {
+                self.scope_tracker.set_unreachable();
+            }
             Operator::CallIndirect { .. } | Operator::Call { .. } => {
                 // add 1 to count the return from the called function
-                self.op_offsets.last_mut().map(|x| *x = 1);
+                self.scope_costs.last_mut().map(|x| *x = 1);
+            }
+            Operator::Block { .. } | Operator::Loop { .. } => {
+                // open a new scope. it will be exactly as reachable as the current scope.
+                self.scope_costs.push(0);
+                self.scope_tracker.open_scope();
+            }
+            Operator::If { .. } => {
+                self.scope_costs.push(0);
+                self.scope_tracker.open_conditional();
             }
-            Operator::Block { .. } | Operator::Loop { .. } | Operator::If { .. } => {
-                // open a new scope
-                self.op_offsets.push(0);
+            Operator::Else => {
+                self.scope_tracker.add_alternate_branch();
```

Comment:
do we need to push onto `scope_costs` here as well? it looks odd that this is the only case that doesn't have a matching pair of `scope_costs` and `scope_tracker` manipulation

Reply:
we do not, but only because `Else` causes the counter to be flushed, so we know it will be 0 entering the body of `Else`'s block. I'll note that as why there's nothing with `scope_costs` here.